// renovate/dist/manager/pip_setup/extract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractPackageFile = exports.extractSetupFile = exports.getPythonAlias = exports.parsePythonVersion = exports.resetModule = exports.pythonVersions = void 0;
const global_1 = require("../../config/global");
const pypi_1 = require("../../datasource/pypi");
const logger_1 = require("../../logger");
const types_1 = require("../../types");
const exec_1 = require("../../util/exec");
const ignore_1 = require("../../util/ignore");
const extract_1 = require("../pip_requirements/extract");
const util_1 = require("./util");
exports.pythonVersions = ['python', 'python3', 'python3.8', 'python3.9'];
let pythonAlias = null;
function resetModule() {
    pythonAlias = null;
}
exports.resetModule = resetModule;
function parsePythonVersion(str) {
    const arr = str.split(' ')[1].split('.');
    return [parseInt(arr[0], 10), parseInt(arr[1], 10)];
}
exports.parsePythonVersion = parsePythonVersion;
async function getPythonAlias() {
    if (pythonAlias) {
        return pythonAlias;
    }
    pythonAlias = exports.pythonVersions[0]; // fallback to 'python'
    for (const pythonVersion of exports.pythonVersions) {
        try {
            const { stdout, stderr } = await (0, exec_1.exec)(`${pythonVersion} --version`);
            const version = parsePythonVersion(stdout || stderr);
            if (version[0] >= 3 && version[1] >= 7) {
                pythonAlias = pythonVersion;
                break;
            }
        }
        catch (err) {
            logger_1.logger.debug(`${pythonVersion} alias not found`);
        }
    }
    return pythonAlias;
}
exports.getPythonAlias = getPythonAlias;
async function extractSetupFile(_content, packageFile, config) {
    let cmd = 'python';
    const extractPy = await (0, util_1.getExtractFile)();
    const args = [`"${extractPy}"`, `"${packageFile}"`];
    if ((0, global_1.getGlobalConfig)().binarySource !== 'docker') {
        logger_1.logger.debug('Running python via global command');
        cmd = await getPythonAlias();
    }
    logger_1.logger.debug({ cmd, args }, 'python command');
    const res = await (0, exec_1.exec)(`${cmd} ${args.join(' ')}`, {
        cwdFile: packageFile,
        timeout: 30000,
        docker: {
            image: 'python',
        },
    });
    if (res.stderr) {
        const stderr = res.stderr
            .replace(/.*\n\s*import imp/, '')
            .trim()
            .replace('fatal: No names found, cannot describe anything.', '');
        if (stderr.length) {
            logger_1.logger.warn({ stdout: res.stdout, stderr }, 'Error in read setup file');
        }
    }
    return (0, util_1.parseReport)(packageFile);
}
exports.extractSetupFile = extractSetupFile;
async function extractPackageFile(content, packageFile, config) {
    logger_1.logger.debug('pip_setup.extractPackageFile()');
    let setup;
    try {
        setup = await extractSetupFile(content, packageFile, config);
    }
    catch (err) {
        logger_1.logger.debug({ err, content, packageFile }, 'Failed to read setup.py file');
    }
    if (!setup) {
        return null;
    }
    const requires = [];
    if (setup.install_requires) {
        requires.push(...setup.install_requires);
    }
    if (setup.extras_require) {
        for (const req of Object.values(setup.extras_require)) {
            requires.push(...req);
        }
    }
    const regex = new RegExp(`^${extract_1.dependencyPattern}`);
    const lines = content.split('\n');
    const deps = requires
        .map((req) => {
        const lineNumber = lines.findIndex((l) => l.includes(req));
        if (lineNumber === -1) {
            return null;
        }
        const rawline = lines[lineNumber];
        let dep = {};
        const [, comment] = rawline.split('#').map((part) => part.trim());
        if ((0, ignore_1.isSkipComment)(comment)) {
            dep.skipReason = types_1.SkipReason.Ignored;
        }
        regex.lastIndex = 0;
        const matches = regex.exec(req);
        if (!matches) {
            return null;
        }
        const [, depName, , currentValue] = matches;
        dep = {
            ...dep,
            depName,
            currentValue,
            managerData: { lineNumber },
            datasource: pypi_1.PypiDatasource.id,
        };
        return dep;
    })
        .filter(Boolean)
        .sort((a, b) => a.managerData.lineNumber === b.managerData.lineNumber
        ? a.depName.localeCompare(b.depName)
        : a.managerData.lineNumber - b.managerData.lineNumber);
    if (!deps.length) {
        return null;
    }
    return { deps };
}
exports.extractPackageFile = extractPackageFile;
