// renovate/dist/platform/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initPlatform = exports.setPlatformApi = exports.platform = exports.getPlatforms = exports.getPlatformList = void 0;
const url_1 = __importDefault(require("url"));
const error_messages_1 = require("../constants/error-messages");
const logger_1 = require("../logger");
const git_1 = require("../util/git");
const hostRules = __importStar(require("../util/host-rules"));
const api_1 = __importDefault(require("./api"));
__exportStar(require("./types"), exports);
const getPlatformList = () => Array.from(api_1.default.keys());
exports.getPlatformList = getPlatformList;
const getPlatforms = () => api_1.default;
exports.getPlatforms = getPlatforms;
// eslint-disable-next-line @typescript-eslint/naming-convention
let _platform;
const handler = {
    get(_target, prop) {
        if (!_platform) {
            throw new Error(error_messages_1.PLATFORM_NOT_FOUND);
        }
        return _platform[prop];
    },
};
exports.platform = new Proxy({}, handler);
function setPlatformApi(name) {
    if (!api_1.default.has(name)) {
        throw new Error(`Init: Platform "${name}" not found. Must be one of: ${(0, exports.getPlatformList)().join(', ')}`);
    }
    _platform = api_1.default.get(name);
}
exports.setPlatformApi = setPlatformApi;
async function initPlatform(config) {
    var _a;
    (0, git_1.setPrivateKey)(config.gitPrivateKey);
    (0, git_1.setNoVerify)((_a = config.gitNoVerify) !== null && _a !== void 0 ? _a : []);
    setPlatformApi(config.platform);
    // TODO: types
    const platformInfo = await exports.platform.initPlatform(config);
    const returnConfig = { ...config, ...platformInfo };
    // istanbul ignore else
    if (config === null || config === void 0 ? void 0 : config.gitAuthor) {
        logger_1.logger.debug(`Using configured gitAuthor (${config.gitAuthor})`);
        returnConfig.gitAuthor = config.gitAuthor;
    }
    else if (platformInfo === null || platformInfo === void 0 ? void 0 : platformInfo.gitAuthor) {
        logger_1.logger.debug(`Using platform gitAuthor: ${String(platformInfo.gitAuthor)}`);
        returnConfig.gitAuthor = platformInfo.gitAuthor;
    }
    // This is done for validation and will be overridden later once repo config is incorporated
    (0, git_1.setGitAuthor)(returnConfig.gitAuthor);
    const platformRule = {
        matchHost: url_1.default.parse(returnConfig.endpoint).hostname,
    };
    ['token', 'username', 'password'].forEach((field) => {
        if (config[field]) {
            platformRule[field] = config[field];
            delete returnConfig[field];
        }
    });
    returnConfig.hostRules = returnConfig.hostRules || [];
    returnConfig.hostRules.push(platformRule);
    hostRules.add(platformRule);
    const typedPlatformRule = {
        ...platformRule,
        hostType: returnConfig.platform,
    };
    returnConfig.hostRules.push(typedPlatformRule);
    hostRules.add(typedPlatformRule);
    return returnConfig;
}
exports.initPlatform = initPlatform;
