// renovate/dist/datasource/adoptium-java/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdoptiumJavaDatasource = void 0;
const external_host_error_1 = require("../../types/errors/external-host-error");
const decorator_1 = require("../../util/cache/package/decorator");
const types_1 = require("../../util/http/types");
const datasource_1 = require("../datasource");
const common_1 = require("./common");
class AdoptiumJavaDatasource extends datasource_1.Datasource {
    constructor() {
        super(common_1.datasource);
        this.customRegistrySupport = false;
        this.defaultRegistryUrls = [common_1.defaultRegistryUrl];
        this.caching = true;
    }
    async getReleases({ registryUrl, }) {
        var _a, _b;
        let page = 0;
        const url = `${registryUrl}v3/info/release_versions?page_size=${common_1.pageSize}&project=jdk&release_type=ga&sort_method=DATE&sort_order=DESC&vendor=adoptium`;
        const result = {
            homepage: 'https://adoptium.net',
            releases: [],
        };
        let resp;
        try {
            do {
                resp = (await this.http.getJson(`${url}&page=${page}`)).body;
                result.releases.push(...resp.versions.map(({ semver }) => ({ version: semver })));
                page += 1;
            } while (page < 50 && resp.versions.length === common_1.pageSize);
        }
        catch (err) {
            // istanbul ignore else: not testable with nock
            if (err instanceof types_1.HttpError) {
                if (((_a = err.response) === null || _a === void 0 ? void 0 : _a.statusCode) === 404 && page > 0) {
                    // no more pages
                    return result;
                }
                if (((_b = err.response) === null || _b === void 0 ? void 0 : _b.statusCode) !== 404) {
                    throw new external_host_error_1.ExternalHostError(err);
                }
            }
            this.handleGenericErrors(err);
        }
        return result.releases.length ? result : null;
    }
}
AdoptiumJavaDatasource.id = common_1.datasource;
__decorate([
    (0, decorator_1.cache)({
        namespace: `datasource-${common_1.datasource}`,
        key: ({ registryUrl }) => `${registryUrl}`,
    })
], AdoptiumJavaDatasource.prototype, "getReleases", null);
exports.AdoptiumJavaDatasource = AdoptiumJavaDatasource;
