// renovate/dist/datasource/ruby-version/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RubyVersionDatasource = void 0;
const external_host_error_1 = require("../../types/errors/external-host-error");
const decorator_1 = require("../../util/cache/package/decorator");
const html_1 = require("../../util/html");
const ruby_1 = require("../../versioning/ruby");
const datasource_1 = require("../datasource");
class RubyVersionDatasource extends datasource_1.Datasource {
    constructor() {
        super(RubyVersionDatasource.id);
        this.defaultRegistryUrls = ['https://www.ruby-lang.org/'];
        this.customRegistrySupport = false;
        this.defaultVersioning = ruby_1.id;
    }
    async getReleases({ registryUrl, }) {
        const res = {
            homepage: 'https://www.ruby-lang.org',
            sourceUrl: 'https://github.com/ruby/ruby',
            releases: [],
        };
        const rubyVersionsUrl = `${registryUrl}en/downloads/releases/`;
        try {
            const response = await this.http.get(rubyVersionsUrl);
            const root = (0, html_1.parse)(response.body);
            const rows = root.querySelector('.release-list').querySelectorAll('tr');
            rows.forEach((row) => {
                const tds = row.querySelectorAll('td');
                const columns = [];
                tds.forEach((td) => columns.push(td.innerHTML));
                if (columns.length) {
                    const version = columns[0].replace('Ruby ', '');
                    if ((0, ruby_1.isVersion)(version)) {
                        const releaseTimestamp = columns[1];
                        const changelogUrl = columns[2]
                            .replace('<a href="', 'https://www.ruby-lang.org')
                            .replace('">more...</a>', '');
                        res.releases.push({ version, releaseTimestamp, changelogUrl });
                    }
                }
            });
        }
        catch (err) {
            this.handleGenericErrors(err);
        }
        return res;
    }
    // eslint-disable-next-line class-methods-use-this
    handleSpecificErrors(err) {
        throw new external_host_error_1.ExternalHostError(err);
    }
}
RubyVersionDatasource.id = 'ruby-version';
__decorate([
    (0, decorator_1.cache)({ namespace: `datasource-${RubyVersionDatasource.id}`, key: 'all' })
], RubyVersionDatasource.prototype, "getReleases", null);
exports.RubyVersionDatasource = RubyVersionDatasource;
