// renovate/dist/datasource/npm/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.customRegistrySupport = exports.defaultVersioning = exports.id = exports.setNpmrc = exports.getNpmrc = exports.getReleases = exports.resetCache = exports.resetMemCache = void 0;
const npmVersioning = __importStar(require("../../versioning/npm"));
var get_1 = require("./get");
Object.defineProperty(exports, "resetMemCache", { enumerable: true, get: function () { return get_1.resetMemCache; } });
Object.defineProperty(exports, "resetCache", { enumerable: true, get: function () { return get_1.resetCache; } });
var releases_1 = require("./releases");
Object.defineProperty(exports, "getReleases", { enumerable: true, get: function () { return releases_1.getReleases; } });
var npmrc_1 = require("./npmrc");
Object.defineProperty(exports, "getNpmrc", { enumerable: true, get: function () { return npmrc_1.getNpmrc; } });
Object.defineProperty(exports, "setNpmrc", { enumerable: true, get: function () { return npmrc_1.setNpmrc; } });
var common_1 = require("./common");
Object.defineProperty(exports, "id", { enumerable: true, get: function () { return common_1.id; } });
exports.defaultVersioning = npmVersioning.id;
exports.customRegistrySupport = false;
