// renovate/dist/manager/github-actions/extract.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractPackageFile = void 0;
const githubTagsDatasource = __importStar(require("../../datasource/github-tags"));
const logger_1 = require("../../logger");
const types_1 = require("../../types");
const dockerVersioning = __importStar(require("../../versioning/docker"));
const extract_1 = require("../dockerfile/extract");
const dockerRe = /^\s+uses: docker:\/\/([^"]+)\s*$/;
const actionRe = /^\s+-?\s+?uses: (?<replaceString>(?<depName>[\w-]+\/[\w-]+)(?<path>\/.*)?@(?<currentValue>.+?)(?: # renovate: tag=(?<tag>.+?))?)\s*?$/;
// SHA1 or SHA256, see https://github.blog/2020-10-19-git-2-29-released/
const shaRe = /^[a-z0-9]{40}|[a-z0-9]{64}$/;
function extractPackageFile(content) {
    logger_1.logger.trace('github-actions.extractPackageFile()');
    const deps = [];
    for (const line of content.split('\n')) {
        if (line.trim().startsWith('#')) {
            continue; // eslint-disable-line no-continue
        }
        const dockerMatch = dockerRe.exec(line);
        if (dockerMatch) {
            const [, currentFrom] = dockerMatch;
            const dep = (0, extract_1.getDep)(currentFrom);
            dep.depType = 'docker';
            dep.versioning = dockerVersioning.id;
            deps.push(dep);
            continue; // eslint-disable-line no-continue
        }
        const tagMatch = actionRe.exec(line);
        if (tagMatch === null || tagMatch === void 0 ? void 0 : tagMatch.groups) {
            const { depName, currentValue, path = '', tag, replaceString, } = tagMatch.groups;
            const dep = {
                depName,
                commitMessageTopic: '{{{depName}}} action',
                datasource: githubTagsDatasource.id,
                versioning: dockerVersioning.id,
                depType: 'action',
                replaceString,
                autoReplaceStringTemplate: `{{depName}}${path}@{{#if newDigest}}{{newDigest}}{{#if newValue}} # renovate: tag={{newValue}}{{/if}}{{/if}}{{#unless newDigest}}{{newValue}}{{/unless}}`,
            };
            if (shaRe.test(currentValue)) {
                dep.currentValue = tag;
                dep.currentDigest = currentValue;
            }
            else {
                dep.currentValue = currentValue;
                if (!dockerVersioning.api.isValid(currentValue)) {
                    dep.skipReason = types_1.SkipReason.InvalidVersion;
                }
            }
            deps.push(dep);
        }
    }
    if (!deps.length) {
        return null;
    }
    return { deps };
}
exports.extractPackageFile = extractPackageFile;
