// renovate/dist/manager/dockerfile/extract.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractPackageFile = exports.getDep = exports.splitImageParts = void 0;
const is_1 = __importDefault(require("@sindresorhus/is"));
const datasourceDocker = __importStar(require("../../datasource/docker"));
const logger_1 = require("../../logger");
const types_1 = require("../../types");
const ubuntuVersioning = __importStar(require("../../versioning/ubuntu"));
function splitImageParts(currentFrom) {
    if (currentFrom.includes('$')) {
        return {
            skipReason: types_1.SkipReason.ContainsVariable,
        };
    }
    const [currentDepTag, currentDigest] = currentFrom.split('@');
    const depTagSplit = currentDepTag.split(':');
    let depName;
    let currentValue;
    if (depTagSplit.length === 1 ||
        depTagSplit[depTagSplit.length - 1].includes('/')) {
        depName = currentDepTag;
    }
    else {
        currentValue = depTagSplit.pop();
        depName = depTagSplit.join(':');
    }
    const dep = {
        depName,
        currentValue,
        currentDigest,
    };
    return dep;
}
exports.splitImageParts = splitImageParts;
const quayRegex = /^quay\.io(?::[1-9][0-9]{0,4})?/i;
function getDep(currentFrom, specifyReplaceString = true) {
    if (!is_1.default.string(currentFrom)) {
        return {
            skipReason: types_1.SkipReason.InvalidValue,
        };
    }
    const dep = splitImageParts(currentFrom);
    if (specifyReplaceString) {
        dep.replaceString = currentFrom;
        dep.autoReplaceStringTemplate =
            '{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}';
    }
    dep.datasource = datasourceDocker.id;
    // Pretty up special prefixes
    if (dep.depName) {
        const specialPrefixes = ['amd64', 'arm64', 'library'];
        for (const prefix of specialPrefixes) {
            if (dep.depName.startsWith(`${prefix}/`)) {
                dep.lookupName = dep.depName;
                dep.depName = dep.depName.replace(`${prefix}/`, '');
                if (specifyReplaceString) {
                    dep.autoReplaceStringTemplate =
                        '{{lookupName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}';
                }
            }
        }
    }
    if (dep.depName === 'ubuntu') {
        dep.versioning = ubuntuVersioning.id;
    }
    // Don't display quay.io ports
    if (quayRegex.test(dep.depName)) {
        const depName = dep.depName.replace(quayRegex, 'quay.io');
        if (depName !== dep.depName) {
            dep.lookupName = dep.depName;
            dep.depName = depName;
            dep.autoReplaceStringTemplate =
                '{{lookupName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}';
        }
    }
    return dep;
}
exports.getDep = getDep;
function extractPackageFile(content) {
    const deps = [];
    const stageNames = [];
    const fromMatches = content.matchAll(/^[ \t]*FROM(?:\\\r?\n| |\t|#.*?\r?\n|[ \t]--[a-z]+=\w+?)*[ \t](?<image>\S+)(?:(?:\\\r?\n| |\t|#.*\r?\n)+as[ \t]+(?<name>\S+))?/gim);
    for (const fromMatch of fromMatches) {
        if (fromMatch.groups.name) {
            logger_1.logger.debug('Found a multistage build stage name');
            stageNames.push(fromMatch.groups.name);
        }
        if (fromMatch.groups.image === 'scratch') {
            logger_1.logger.debug('Skipping scratch');
        }
        else if (stageNames.includes(fromMatch.groups.image)) {
            logger_1.logger.debug({ image: fromMatch.groups.image }, 'Skipping alias FROM');
        }
        else {
            const dep = getDep(fromMatch.groups.image);
            logger_1.logger.trace({
                depName: dep.depName,
                currentValue: dep.currentValue,
                currentDigest: dep.currentDigest,
            }, 'Dockerfile FROM');
            deps.push(dep);
        }
    }
    const copyFromMatches = content.matchAll(/^[ \t]*COPY(?:\\\r?\n| |\t|#.*\r?\n|[ \t]--[a-z]+=\w+?)*[ \t]--from=(?<image>\S+)/gim);
    for (const copyFromMatch of copyFromMatches) {
        if (stageNames.includes(copyFromMatch.groups.image)) {
            logger_1.logger.debug({ image: copyFromMatch.groups.image }, 'Skipping alias COPY --from');
        }
        else if (Number.isNaN(Number(copyFromMatch.groups.image))) {
            const dep = getDep(copyFromMatch.groups.image);
            logger_1.logger.debug({
                depName: dep.depName,
                currentValue: dep.currentValue,
                currentDigest: dep.currentDigest,
            }, 'Dockerfile COPY --from');
            deps.push(dep);
        }
        else {
            logger_1.logger.debug({ image: copyFromMatch.groups.image }, 'Skipping index reference COPY --from');
        }
    }
    if (!deps.length) {
        return null;
    }
    for (const d of deps) {
        d.depType = 'stage';
    }
    deps[deps.length - 1].depType = 'final';
    return { deps };
}
exports.extractPackageFile = extractPackageFile;
