// renovate/dist/manager/gradle/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultConfig = exports.language = exports.updateDependency = exports.extractAllPackageFiles = void 0;
const languages_1 = require("../../constants/languages");
const gradleVersioning = __importStar(require("../../versioning/gradle"));
const deep = __importStar(require("./deep"));
const shallow = __importStar(require("./shallow"));
function extractAllPackageFiles(config, packageFiles) {
    return config.deepExtract
        ? deep.extractAllPackageFiles(config, packageFiles)
        : shallow.extractAllPackageFiles(config, packageFiles);
}
exports.extractAllPackageFiles = extractAllPackageFiles;
function updateDependency(params) {
    var _a;
    return ((_a = params.upgrade) === null || _a === void 0 ? void 0 : _a.deepExtract)
        ? deep.updateDependency(params)
        : shallow.updateDependency(params);
}
exports.updateDependency = updateDependency;
exports.language = languages_1.LANGUAGE_JAVA;
exports.defaultConfig = {
    fileMatch: [
        '\\.gradle(\\.kts)?$',
        '(^|/)gradle.properties$',
        '(^|\\/)gradle\\/.+\\.toml$',
        '\\.versions\\.toml$',
    ],
    timeout: 600,
    versioning: gradleVersioning.id,
};
