// renovate/dist/manager/composer/artifacts.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateArtifacts = void 0;
const is_1 = __importDefault(require("@sindresorhus/is"));
const shlex_1 = require("shlex");
const global_1 = require("../../config/global");
const error_messages_1 = require("../../constants/error-messages");
const platforms_1 = require("../../constants/platforms");
const datasourcePackagist = __importStar(require("../../datasource/packagist"));
const logger_1 = require("../../logger");
const exec_1 = require("../../util/exec");
const fs_1 = require("../../util/fs");
const git_1 = require("../../util/git");
const hostRules = __importStar(require("../../util/host-rules"));
const utils_1 = require("./utils");
function getAuthJson() {
    var _a, _b;
    const authJson = {};
    const githubCredentials = hostRules.find({
        hostType: platforms_1.PLATFORM_TYPE_GITHUB,
        url: 'https://api.github.com/',
    });
    if (githubCredentials === null || githubCredentials === void 0 ? void 0 : githubCredentials.token) {
        authJson['github-oauth'] = {
            'github.com': githubCredentials.token.replace('x-access-token:', ''),
        };
    }
    (_a = hostRules
        .findAll({ hostType: platforms_1.PLATFORM_TYPE_GITLAB })) === null || _a === void 0 ? void 0 : _a.forEach((gitlabHostRule) => {
        if (gitlabHostRule === null || gitlabHostRule === void 0 ? void 0 : gitlabHostRule.token) {
            const host = gitlabHostRule.resolvedHost || 'gitlab.com';
            authJson['gitlab-token'] = authJson['gitlab-token'] || {};
            authJson['gitlab-token'][host] = gitlabHostRule.token;
            // https://getcomposer.org/doc/articles/authentication-for-private-packages.md#gitlab-token
            authJson['gitlab-domains'] = [
                host,
                ...(authJson['gitlab-domains'] || []),
            ];
        }
    });
    (_b = hostRules
        .findAll({ hostType: datasourcePackagist.id })) === null || _b === void 0 ? void 0 : _b.forEach((hostRule) => {
        const { resolvedHost, username, password } = hostRule;
        if (resolvedHost && username && password) {
            authJson['http-basic'] = authJson['http-basic'] || {};
            authJson['http-basic'][resolvedHost] = { username, password };
        }
    });
    return is_1.default.emptyObject(authJson) ? null : JSON.stringify(authJson);
}
async function updateArtifacts({ packageFileName, updatedDeps, newPackageFileContent, config, }) {
    var _a, _b;
    logger_1.logger.debug(`composer.updateArtifacts(${packageFileName})`);
    const lockFileName = packageFileName.replace(/\.json$/, '.lock');
    const existingLockFileContent = await (0, fs_1.readLocalFile)(lockFileName, 'utf8');
    if (!existingLockFileContent) {
        logger_1.logger.debug('No composer.lock found');
        return null;
    }
    const vendorDir = (0, fs_1.getSiblingFileName)(packageFileName, 'vendor');
    const commitVendorFiles = await (0, fs_1.localPathExists)(vendorDir);
    await (0, fs_1.ensureLocalDir)(vendorDir);
    try {
        await (0, fs_1.writeLocalFile)(packageFileName, newPackageFileContent);
        const constraints = {
            ...(0, utils_1.extractContraints)(JSON.parse(newPackageFileContent), JSON.parse(existingLockFileContent)),
            ...config.constraints,
        };
        if (config.isLockFileMaintenance) {
            await (0, fs_1.deleteLocalFile)(lockFileName);
        }
        const preCommands = [
            `install-tool composer ${await (0, utils_1.getComposerConstraint)(constraints)}`,
        ];
        const execOptions = {
            cwdFile: packageFileName,
            extraEnv: {
                COMPOSER_CACHE_DIR: await (0, fs_1.ensureCacheDir)('composer'),
                COMPOSER_AUTH: getAuthJson(),
            },
            docker: {
                preCommands,
                image: 'php',
                tagConstraint: (0, utils_1.getPhpConstraint)(constraints),
                tagScheme: utils_1.composerVersioningId,
            },
        };
        const cmd = 'composer';
        let args;
        if (config.isLockFileMaintenance) {
            args = 'install';
        }
        else {
            args =
                ('update ' + updatedDeps.map((dep) => (0, shlex_1.quote)(dep.depName)).join(' ')).trim() + ' --with-dependencies';
        }
        if (config.composerIgnorePlatformReqs) {
            if (config.composerIgnorePlatformReqs.length === 0) {
                args += ' --ignore-platform-reqs';
            }
            else {
                config.composerIgnorePlatformReqs.forEach((req) => {
                    args += ' --ignore-platform-req ' + (0, shlex_1.quote)(req);
                });
            }
        }
        args += ' --no-ansi --no-interaction';
        if (!(0, global_1.getGlobalConfig)().allowScripts || config.ignoreScripts) {
            args += ' --no-scripts --no-autoloader --no-plugins';
        }
        logger_1.logger.debug({ cmd, args }, 'composer command');
        await (0, exec_1.exec)(`${cmd} ${args}`, execOptions);
        const status = await (0, git_1.getRepoStatus)();
        if (!status.modified.includes(lockFileName)) {
            return null;
        }
        logger_1.logger.debug('Returning updated composer.lock');
        const res = [
            {
                file: {
                    name: lockFileName,
                    contents: await (0, fs_1.readLocalFile)(lockFileName),
                },
            },
        ];
        if (!commitVendorFiles) {
            return res;
        }
        logger_1.logger.debug(`Commiting vendor files in ${vendorDir}`);
        for (const f of [...status.modified, ...status.not_added]) {
            if (f.startsWith(vendorDir)) {
                res.push({
                    file: {
                        name: f,
                        contents: await (0, fs_1.readLocalFile)(f),
                    },
                });
            }
        }
        for (const f of status.deleted) {
            res.push({
                file: {
                    name: '|delete|',
                    contents: f,
                },
            });
        }
        return res;
    }
    catch (err) {
        // istanbul ignore if
        if (err.message === error_messages_1.TEMPORARY_ERROR) {
            throw err;
        }
        if ((_a = err.message) === null || _a === void 0 ? void 0 : _a.includes('Your requirements could not be resolved to an installable set of packages.')) {
            logger_1.logger.info('Composer requirements cannot be resolved');
        }
        else if ((_b = err.message) === null || _b === void 0 ? void 0 : _b.includes('write error (disk full?)')) {
            throw new Error(error_messages_1.SYSTEM_INSUFFICIENT_DISK_SPACE);
        }
        else {
            logger_1.logger.debug({ err }, 'Failed to generate composer.lock');
        }
        return [
            {
                artifactError: {
                    lockFile: lockFileName,
                    stderr: err.message,
                },
            },
        ];
    }
}
exports.updateArtifacts = updateArtifacts;
