// renovate/dist/manager/argocd/extract.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractPackageFile = void 0;
const js_yaml_1 = require("js-yaml");
const gitTags = __importStar(require("../../datasource/git-tags"));
const helm_1 = require("../../datasource/helm");
const util_1 = require("./util");
function createDependency(definition) {
    var _a;
    const source = (_a = definition.spec) === null || _a === void 0 ? void 0 : _a.source;
    if (source == null) {
        return null;
    }
    // a chart variable is defined this is helm declaration
    if (source.chart) {
        return {
            depName: source.chart,
            registryUrls: [source.repoURL],
            currentValue: source.targetRevision,
            datasource: helm_1.HelmDatasource.id,
        };
    }
    return {
        depName: source.repoURL,
        currentValue: source.targetRevision,
        datasource: gitTags.id,
    };
}
function extractPackageFile(content, fileName, config) {
    // check for argo reference. API version for the kind attribute is used
    if (util_1.fileTestRegex.test(content) === false) {
        return null;
    }
    const definitions = (0, js_yaml_1.loadAll)(content);
    const deps = definitions
        .map((definition) => createDependency(definition))
        .filter(Boolean);
    return deps.length ? { deps } : null;
}
exports.extractPackageFile = extractPackageFile;
