// renovate/dist/workers/pr/body/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPrBody = void 0;
const platform_1 = require("../../../platform");
const template = __importStar(require("../../../util/template"));
const changelogs_1 = require("./changelogs");
const config_description_1 = require("./config-description");
const controls_1 = require("./controls");
const footer_1 = require("./footer");
const header_1 = require("./header");
const notes_1 = require("./notes");
const updates_table_1 = require("./updates-table");
function massageUpdateMetadata(config) {
    config.upgrades.forEach((upgrade) => {
        /* eslint-disable no-param-reassign */
        const { homepage, sourceUrl, sourceDirectory, changelogUrl, dependencyUrl, } = upgrade;
        let depNameLinked = upgrade.depName;
        const primaryLink = homepage || sourceUrl || dependencyUrl;
        if (primaryLink) {
            depNameLinked = `[${depNameLinked}](${primaryLink})`;
        }
        const otherLinks = [];
        if (homepage && sourceUrl) {
            otherLinks.push(`[source](${sourceUrl})`);
        }
        if (changelogUrl) {
            otherLinks.push(`[changelog](${changelogUrl})`);
        }
        if (otherLinks.length) {
            depNameLinked += ` (${otherLinks.join(', ')})`;
        }
        upgrade.depNameLinked = depNameLinked;
        const references = [];
        if (homepage) {
            references.push(`[homepage](${homepage})`);
        }
        if (sourceUrl) {
            let fullUrl = sourceUrl;
            if (sourceDirectory) {
                fullUrl =
                    sourceUrl.replace(/\/?$/, '/') +
                        'tree/HEAD/' +
                        sourceDirectory.replace('^/?/', '');
            }
            references.push(`[source](${fullUrl})`);
        }
        if (changelogUrl) {
            references.push(`[changelog](${changelogUrl})`);
        }
        upgrade.references = references.join(', ');
        /* eslint-enable no-param-reassign */
    });
}
async function getPrBody(config) {
    massageUpdateMetadata(config);
    const content = {
        header: (0, header_1.getPrHeader)(config),
        table: (0, updates_table_1.getPrUpdatesTable)(config),
        notes: (0, notes_1.getPrNotes)(config) + (0, notes_1.getPrExtraNotes)(config),
        changelogs: (0, changelogs_1.getChangelogs)(config),
        configDescription: await (0, config_description_1.getPrConfigDescription)(config),
        controls: await (0, controls_1.getControls)(config),
        footer: (0, footer_1.getPrFooter)(config),
    };
    const prBodyTemplate = config.prBodyTemplate;
    let prBody = template.compile(prBodyTemplate, content, false);
    prBody = prBody.trim();
    prBody = prBody.replace(/\n\n\n+/g, '\n\n');
    prBody = platform_1.platform.massageMarkdown(prBody);
    return prBody;
}
exports.getPrBody = getPrBody;
